### PR TITLE
Completed gmwm.robust and plot.auto.imu with decomposed processes

### PR DESCRIPTION
**Pull request by Wenchao-Yang**
Also wrote the function to check the parameters
Issue #144 and #119 are done.

### CODE REVIEW

**Comment by Wenchao-Yang** [comment]
Travis-ci gives one note:
```
checking installed package size ... NOTE
  installed size is 21.0Mb
  sub-directories of 1Mb or more:
 libs  19.1Mb 
```
And one error:
```
checking examples ... ERROR
Running examples in ‘gmwm-Ex.R’ failed
The error most likely occurred in: 
> set.seed(8836)
> x = gen.gts(AR1(phi = .1, sigma2 = 1) + AR1(phi = 0.95, sigma2 = .1), 1000)
> obj = gmwm.robust(2*AR1()+WN(), data = x)
> compare.eff(obj)
Error in unit(x, default.units) : 'x' and 'units' must have length > 0
Calls: <Anonymous> ... <Anonymous> -> f -> ggname -> grobName -> polygonGrob -> unit
Execution halted DONE
```
This is one example I wrote in gmwm.robust(). It seems to be related to one bug in ggplot2:
https://github.com/hadley/ggplot2/issues/905

I didn't see this error and note on my computer, and the check succeeded completely.

@coatless Any idea how to fix it?

**Comment by Wenchao-Yang** [comment]
Also, I don't think compare.eff draws points outside the bound, because I didn't set the limit in the function.

**Comment by coatless** [comment]
@Wenchao-Yang 

The `compare.eff()` function is DOA. You will need to go back and fix it. The issue is not with hadley/ggplot2#905.

That is, the function is not working under even normal operating conditions. e.g.

```r
set.seed(8836)
x = gen.gts(AR1(phi = .1, sigma2 = 1) + AR1(phi = 0.95, sigma2 = .1), 1000)
obj_classical = gmwm(2*AR1()+WN(), data = x)
obj_robust = gmwm(2*AR1()+WN(), data = x, robust = T)
compare.eff(obj_classical,obj_robust)
```

Thus, I think something is not being assigned correctly.


Similarly, running the manual example outside of travis yields  the same error. 

e.g.

```r
set.seed(8836)
x = gen.gts(AR1(phi = .1, sigma2 = 1) + AR1(phi = 0.95, sigma2 = .1), 1000)
obj = gmwm.robust(2*AR1()+WN(), data = x)
compare.eff(obj)
```

Error:

> Error in unit(x, default.units) : 'x' and 'units' must have length > 0

Are the objects correctly being evaluated within the function? It seems like a term might be null.

**Comment by coatless** [comment]
Addresses one issue in #150 as it relates to 

* Move out parameter checks into their own callable function.

**Comment by Wenchao-Yang** [comment]
@coatless 
It's a bit weird because the examples you provided work perfectly well on my computer. Let's talk about this tomorrow.

File: R/misc.R [hunk 1]
```diff
@@ -254,3 +254,75 @@ getEff = function(obj.list){
...
+#' @keywords internal
+#' @details 
+#' 
+#' The user supplied parameters are usually \code{line.color}, \code{line.type}, \code{point.size}, 
+#' \code{point.shape}, \code{CI.color} and \code{legend.label}.
+#' 
+#' This function will check whether the required length of the parameter is met. If not, it will assign the 
+#' default value to that parameter.
+#' 
+checkParams = function(params, require.len, default, null.is.fine, env = parent.frame()){
```
Comment (Wenchao-Yang) [inline R/misc.R]: This is how I implemented the parameter checking function.

File: R/GMWM.R [hunk 2]
```diff
@@ -2186,17 +2199,8 @@ compare.eff = function(..., display.eff = T, order.eff = T, facet.label = NULL,
...
-        isNull = is.null(get(one_param))
-        if(isNull && nullIsFine[i]){}else{
-          warning(paste('Parameter', one_param, 'requires', requireLength[i],'elements,','but', length(get(one_param)),
-                        'is supplied.','Default setting is used.'))
-        }
-        assign(one_param, default[[i]])
-      }
-    }
+
+    checkParams(params = params, require.len = requireLength, default = default, null.is.fine = nullIsFine)
```
Comment (Wenchao-Yang) [inline R/GMWM.R]: This is how I use it in compare.eff().

**Comment by coatless** [comment]
@Wenchao-Yang 

Sure. 1 - 2 PM okay? 

Before you swing by, could you try cloning and rebuilding the package off the submitted branch?

e.g. use:

[click me](github-windows://openRepo/https://github.com/SMAC-Group/gmwm?branch=wenchao)

**Comment by Wenchao-Yang** [comment]
Works for me.
I'll do it now.

**Comment by coatless** [comment]
Seems like all the `compare.*()` functions broke under ggplot2 v2.1.0 on R 3.2.4

**Comment by coatless** [comment]
Issue with:

      p = p + facet_grid(dataset_h ~ dataset_v)

**Comment by coatless** [comment]
This issue seems to come from confusion between which data.frame to extract the `dataset_h ~ dataset_v` values.

**Comment by coatless** [comment]
The issue relates to using the `geom_*()` function to clean the data via `na.rm = T`. Clean the data of `NA` before supplying it within the `data = ` using `na.omit()`. Better yet, why are `NA`s entering the `data.frame` at all?

Works:

```r
# GMWM Objects
set.seed(8836)
n = 1000
x = gen.gts(AR1(phi = .1, sigma2 = 1) + AR1(phi = 0.95, sigma2 = .1), n)
x[1] = 1000 #Robust gmwm works better for contaminated data
GMWM1 = gmwm(2*AR1()+RW(), data = x, robust  = TRUE, eff = 0.9)
GMWM2 = gmwm(2*AR1()+RW(), data = x, robust  = TRUE, eff = 0.6)

# Graph data.frame

obj_list = list(GMWM1, GMWM2) 
numObj = length(obj_list)

object.names = sapply(obj_list, FUN = function(x){formatRobustEff(x)} )

# Problem: When object names are the same, function will not work
# Solution: Add space after object names
object.names = addSpaceIfDuplicate(object.names)

# S3: Rearrange the data into a data frame which can be passed to next step
total.len = 0
each.len = numeric(numObj)

for (i in 1:numObj ){
  
  each.len[i] = length(obj_list[[i]]$wv.empir)
  total.len = total.len + each.len[i]*numObj
}

# Choose the column names for empty data frame
nlen = nchar(numObj)
WV_names = sprintf(paste0("WV%0",nlen,"d"),1:numObj)
low_names = sprintf(paste0("low%0",nlen,"d"),1:numObj)
high_names = sprintf(paste0("high%0",nlen,"d"),1:numObj)
z_theo_names = sprintf(paste0("z_theo%0",nlen,"d"),1:numObj)
col.names = c('scales', 'dataset_h', 'dataset_v', 
              WV_names, low_names, high_names, z_theo_names)

# Initialize empty data frame with right number of rows
num.col = 3 + 4*numObj
obj = as.data.frame(matrix(NA, ncol = num.col, nrow = total.len), stringsAsFactors = FALSE)
colnames(obj) = col.names

# Put data into data.frame
t = 1
for (i in 1:numObj){
  for(j in 1:numObj){
    
    ## Diagonal ========================
    if(i == j){
      #compare to itself
      #set some values to NA
      d = each.len[i]
      
      WV_name = WV_names[i]
      low_name = low_names[i] 
      high_name = high_names[i] 
      z_theo_name = z_theo_names[i]
      
      obj[t:(t+d-1),c('scales', 'dataset_h', 'dataset_v', 
                      WV_name, low_name, high_name, z_theo_name)] = 
        
        data.frame(obj_list[[i]]$scales,
                   
                   object.names[i],
                   object.names[j],
                   
                   obj_list[[i]]$wv.empir,
                   obj_list[[i]]$ci.low,
                   obj_list[[i]]$ci.high,
                   obj_list[[i]]$theo, stringsAsFactors = F)
      
      t = t +d
      
    }else if (i > j ){
      ## lower triagular ========================
      
      d = each.len[i]
      
      WV_name = WV_names[c(i,j)]
      low_name = low_names[c(i,j)] 
      high_name = high_names[c(i,j)] 
      #z_theo_name = z_theo_names[c(i,j)]
      
      obj[t:(t+d-1),c('scales', 'dataset_h', 'dataset_v', 
                      WV_name, low_name, high_name)] = 
        
        data.frame(obj_list[[i]]$scales,
                   
                   object.names[i],
                   object.names[j],
                   
                   obj_list[[i]]$wv.empir,
                   obj_list[[j]]$wv.empir,
                   
                   obj_list[[i]]$ci.low,
                   obj_list[[j]]$ci.low,
                   
                   obj_list[[i]]$ci.high,
                   obj_list[[j]]$ci.high, stringsAsFactors = F)
      
      t = t +d
      
    }else{
      ## upper triagular ======================== 
      
      d = each.len[i]
      
      z_theo_name = z_theo_names[c(i,j)]
      
      obj[t:(t+d-1),c('scales','dataset_h','dataset_v',z_theo_name)] = 
        
        data.frame(obj_list[[i]]$scales,
                   
                   object.names[i],
                   object.names[j],
                   
                   obj_list[[i]]$theo,
                   obj_list[[j]]$theo, stringsAsFactors = F)
      
      t = t +d
    }
  } # end of j loop
}# end of i loop

# Change the order of facet label
obj$dataset_h = factor(obj$dataset_h, levels = object.names ) # change the plotting order
obj$dataset_v = factor(obj$dataset_v, levels = object.names )


# Melting the data to long form from wide. 

# Graphs the lines
line_df = melt(obj, id.vars = c('scales', 'dataset_v', 'dataset_h'), factorsAsStrings = F)

# Plots the CI
o1 = obj[,c('scales', 'dataset_h', 'dataset_v', low_names)] 
o2 = obj[,c('scales', 'dataset_h', 'dataset_v', high_names)] 

low_df = melt(o1, measure.vars = low_names, variable.name = 'low', factorsAsStrings = F )
high_df = melt(o2, measure.vars = high_names, variable.name = 'high', factorsAsStrings = F )

colnames(low_df) = c('scales', 'dataset_h', 'dataset_v', 'low', 'l_value')
CI_df =  data.frame(low_df, 
                    high = high_df$high, 
                    h_value = high_df$value) 


# Creating the graph... 

# Remove NAs
line_df = na.omit(line_df)
CI_df = na.omit(CI_df)

# Create the ggplot object
p = ggplot() + 
  geom_line(data = line_df, mapping = aes(x = scales, y = value, color = variable)) + 
  geom_point(data = line_df, mapping = aes(x = scales, y = value, color = variable)) +
  geom_ribbon(data = CI_df, mapping = aes(x = scales, ymin = l_value, ymax = h_value, fill = low), alpha = 0.1) +  
  scale_y_log10(breaks = trans_breaks("log10", function(x) 10^x),
                labels = trans_format("log10", math_format(10^.x))) + 
  scale_x_log10(breaks = trans_breaks("log10", function(x) 10^x),
                labels = trans_format("log10", math_format(10^.x)))


p

# Add facetting
q = p + facet_grid(dataset_h~dataset_v, scales = 'free') 

q

# Check ggplot render build
ggplot_build(q)
```

**Comment by Wenchao-Yang** [comment]
@coatless Thanks for your suggestion. I think it will work!

The NA values are used to draw invisible lines. For example, if you compare two gmwm objects with compare.eff(), although some graphs seem to have 2 lines, they actually have all 6 lines, four of which are invisible (due to the NA values). 

I checked that it would also work if we remove the NA before we draw the graph.

**Comment by coatless** [comment]
@Wenchao-Yang 

The `NA` values enter the `data.frame` by construction as the current setup opts to create a wide `data.frame` initialized with `NA` placeholders  before filling via a `for` loop and then convert it to a long `data.frame` via `melt()`. The melting procedure does not remove the `NA` values unless:

```r
melt(data, na.rm = T)
```

Furthermore, `NA` values by default are dropped by `ggplot` and they would give an error message. Thus, in several of the graphing functions that rely upon `melt()`, there is a considerable amount of `na.rm = T` that has been used to suppress said errors.

Please patch this with `na.rm = T` in `melt()` and update any `compare.*()` functions such as `compare.models()` that suffer from the same problem. 

Thanks!

**Comment by Wenchao-Yang** [comment]
@coatless Everything is done!

**Comment by coatless** [comment]
Awesome. Much appreciated.

**Pull request merged**